### PR TITLE
perf(nuxt): use granular watcher to avoid crawling ignored dirs

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

https://github.com/nuxt/nuxt/issues/12376

### ❓ Type of change
- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This follows up previous focus on watcher improvements (https://github.com/nuxt/nuxt/pull/20179) as a hybrid solution using `chokidar` - it filters out/ignores child directories within `srcDir` directly. In tests in the monorepo, it reduces chokdar startup on a M1 Mac by 50%. I would expect even more significant performance improvements on a windows machine.

To enable this you'll need to set `experimental.watcher` to `granular`. (Otherwise behaviour will stay the same as currently.)

I'd appreciate review - my main concern is whether we ought to make provision for users _deliberately_ seeking to watch changes in `.nuxt` or `node_modules` (which they have enabled with a negative glob pattern in `ignore`). 

cc: @stafyniaksacha

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt/src/core/builder.ts [hunk 1]
```diff
@@ -92,23 +93,41 @@ async function watch (nuxt: Nuxt) {
...
+    watcher.on('all', (event, path) => {
+      if (!pending) {
+        nuxt.callHook('builder:watch', event, normalize(path))
+      }
+      if (event === 'unlinkDir' && path in watchers) {
+        watchers[path].close()
+        delete watchers[path]
+      }
+      if (event === 'addDir' && path !== dir && !ignoredDirs.has(path) && !(path in watchers) && !isIgnored(path)) {
+        watchers[path] = chokidar.watch(path, { ...nuxt.options.watchers.chokidar, ignored: [isIgnored] })
```
Comment (stafyniaksacha) [inline packages/nuxt/src/core/builder.ts:119]: Are the options `ignoreInitial: false, depth: 0` missed voluntary there?

File: packages/schema/src/config/experimental.ts [hunk 2]
```diff
@@ -165,10 +165,13 @@ export default defineUntypedSchema({
...
      * on Windows platforms.
      *
+     * You can also try setting this to `granular` to use an experimental granular
+     * watcher, which ignores top-level directories (like `node_modules` and `.git`).
+     *
      * @see https://github.com/paulmillr/chokidar
      * @see https://github.com/parcel-bundler/watcher
      * @default chokidar
-     * @type {'chokidar' | 'parcel'}
+     * @type {'chokidar' | 'parcel' | 'granular'}
```
Comment (antfu) [inline packages/schema/src/config/experimental.ts:174]:
```suggestion
     * @type {'chokidar' | 'parcel' | 'chokidar-granular'}
```

Do you think it makes sense to do so, indicating it's still chokidar?

**Review by antfu — APPROVED** [review]
Amazing idea! Love the approach!

**Review by atinux — APPROVED** [review]

**Pull request merged**